### PR TITLE
Add Vary: Accept-Encoding if response was compressed

### LINKED ISSUE #187: Compression: Vary Header not returned when content gzipped
Response header should contain
```
Vary: Accept-Encoding
```

when content is gzipped. IIS compression currently adds header.

### PR DESCRIPTION
**Pull request by JunTaoLuo**
Fixes #187

### CODE REVIEW

File: src/Microsoft.AspNetCore.ResponseCompression/BodyWrapperStream.cs [hunk 1]
```diff
@@ -214,6 +214,11 @@ private void OnWrite()
                     var compressionProvider = ResolveCompressionProvider();
                     if (compressionProvider != null)
                     {
+                        if (!HeaderUtilities.ContainsCacheDirective(HeaderNames.Vary, HeaderNames.AcceptEncoding))
+                        {
+                            _context.Response.Headers.Add(HeaderNames.Vary, HeaderNames.AcceptEncoding);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCompression/BodyWrapperStream.cs]: Append or AppendCommaSeperatedValue

File: src/Microsoft.AspNetCore.ResponseCompression/BodyWrapperStream.cs [hunk 2]
```diff
@@ -214,6 +214,11 @@ private void OnWrite()
                     var compressionProvider = ResolveCompressionProvider();
                     if (compressionProvider != null)
                     {
+                        if (!HeaderUtilities.ContainsCacheDirective(HeaderNames.Vary, HeaderNames.AcceptEncoding))
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCompression/BodyWrapperStream.cs]: This utility is named poorly but does what is needed which is to scan the values of the headers and see if the specific value, in this case Accept-Encoding, exist.

Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCompression/BodyWrapperStream.cs]: File a bug for it

Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCompression/BodyWrapperStream.cs]: Actually what I said previously isn't accurate, the logic in HeaderUtilities is specific for cache directives. This is because it only handles the cases where the header values are in the format `valueName=<someValue>, ...` which doesn't generalize to other headers such as Accept where the headers can look like `valueName;q=<qValue>, ...`. For the use case here, the layout of the header value is even simpler since it is a comma separated list of tokens. I'll add a separate helper function.

File: test/Microsoft.AspNetCore.ResponseCompression.Tests/ResponseCompressionMiddlewareTest.cs [hunk 3]
```diff
@@ -793,13 +793,24 @@ private void CheckResponseCompressed(HttpResponseMessage response, int expectedB
         {
             IEnumerable<string> contentMD5 = null;
 
+            var containsVaryAcceptEncoding = false;
+            foreach (var value in response.Headers.GetValues(HeaderNames.Vary))
+            {
+                if (value.Contains(HeaderNames.AcceptEncoding))
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.ResponseCompression.Tests/ResponseCompressionMiddlewareTest.cs]: why a Contains vs. an exact match?

Comment (Tratcher) [inline test/Microsoft.AspNetCore.ResponseCompression.Tests/ResponseCompressionMiddlewareTest.cs]: Or are you trying to avoid parsing the header into multiple segments?

Comment (JunTaoLuo) [inline test/Microsoft.AspNetCore.ResponseCompression.Tests/ResponseCompressionMiddlewareTest.cs]: In this case either would work since the tests only set one value to the Vary header but in general when we are dealing with headers, we can't just assume each string in StringValues contains one value. We ran into this problem when we received Header["Cache-Control"] = new[] {"no-cache,no-store"} from MVC for example. So it's best to check using contains rather than exact match. 

In the same regard, we also shouldn't be assuming the format of the headers. For example, the test could break with a bad Vary header like Headers["Vary"] = new [] {"badValue=Accept-Encoding"} but I ignored this corner case in the tests.

File: test/Microsoft.AspNetCore.ResponseCompression.Tests/BodyWrapperStreamTests.cs [hunk 4]
```diff
@@ -9,12 +9,32 @@
 using Microsoft.AspNetCore.Http;
 using Moq;
 using Xunit;
+using Microsoft.Net.Http.Headers;
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.ResponseCompression.Tests/BodyWrapperStreamTests.cs]: Sort

**Comment by JunTaoLuo** [comment]
Tested with the response caching middleware and filed a bug https://github.com/aspnet/ResponseCaching/issues/92. Currently if the response is compressed before being cached, the expected behaviour where the response is cached per accept encoding is observed. If the compression comes after caching, however, the response is still cached per accept encoding. This issue will be fixed in the response caching middleware.
